src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Identity.Client.Advanced
+{
+    /// <summary>
+    /// </summary>
+    public static class AcquireTokenParameterBuilderExtensions
+    {
+        /// <summary>
+        /// Adds additional Http Headers to the token request.
+        /// </summary>
+        /// <param name="builder">Parameter builder for a acquiring tokens.</param>
+        /// <param name="extraHttpHeaders">additional Http Headers to add to the token request.</param>
+        /// <returns></returns>
+        public static AcquireTokenInteractiveParameterBuilder WithExtraHttpHeaders(
```

Comment:
I tried to use the base parameter builder classes so that I only needed one method here but they are generics and caused a lot of strange build errors. I ended up adding an overload for each of the child parameter builder types.
@bgavrilMS

Reply:
that's not the right solution, @trwalke.

Reply:
Yes, the use generics here is very tricky. In hindsight, we should have opted for a simpler construct that does not rely on abstract generic classes with recursive generics. It also has the downside that it cannot be unit tested :(


Anyway, here's the magic formula, which I happen to stumble upon

```csharp
 /// <summary>
    /// Extensions for <see cref="AbstractAcquireTokenParameterBuilder{T}"/>
    /// </summary>
    public static class AbstractAcquireTokenParameterBuilderExtension
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="builder"></param>
        /// <param name="extraHttpHeaders"></param>
        /// <returns>The builder to chain the .With methods.</returns>
        public static AbstractAcquireTokenParameterBuilder<T> WithExtraHttpHeader<T>(this AbstractAcquireTokenParameterBuilder<T> builder, IDictionary<string, string> extraHttpHeaders)
            where T : AbstractAcquireTokenParameterBuilder<T>
        {
            builder.CommonParameters.ExtraHttpHeaders = extraHttpHeaders;
            return builder;
        }
    }
```